lib/mxcreply.cpp

```diff
@@ -0,0 +1,55 @@
+// SPDX-FileCopyrightText: Tobias Fella <fella@posteo.de>
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "mxcreply.h"
+
+#include <QStandardPaths>
+#include "connection.h"
+#include "jobs/downloadfilejob.h"
+#include "room.h"
+
+using namespace Quotient;
+
+class MxcReply::Private
+{
+public:
+    Quotient::DownloadFileJob *m_job = nullptr;
+    QIODevice* m_buffer;
+};
+
+MxcReply::MxcReply(Room* room, const QString &eventId, const QString& mediaId)
+    : d(std::make_unique<Private>())
+{
+    if(!mediaId.contains("/")) {
+        qWarning() << "Ivalid media id" << mediaId;
+        return;
+    }
+    auto path = QStringLiteral("%1/downloads/%2").arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation), mediaId);
+    QString localPath = QFileInfo(path).absolutePath();
+    QDir dir;
+    if (!dir.exists(localPath)) {
+        dir.mkpath(localPath);
+    }
+    d->m_job = room->connection()->downloadFile(QUrl(mediaId), path);
```

Comment:
This feels wrong. So far throughout the library we have the dependency tree like this:
```
Room -> Connection |
BaseJob            | -> ConnectionData -> NetworkAccessManager
```
I can understand that some room properties might be needed at `NAM` level as a kind of context but creating a job from inside `NAM` produces a fairly deep reverse dependency from `NAM` to `BaseJob`. Since `DownloadFileJob` is relatively straightforward I wouldn't mind if it were collapsed into `NAM`/`MxcReply` instead (dropping `DownloadFileJob` would also allow to simplify `BaseJob` in a couple of places - but `BaseJob` is up for rewriting anyway when I get to futures).